docs: pre-generate address (#1691)
alchemyplatform/aa-sdk: docs/pages/react/pregenerate-wallets.mdx

## Prerequisites

- Alchemy **API Key** with Smart Wallets enabled
  - If starting from scratch, follow the [quickstart](/docs/wallets/react/quickstart/existing-project#2-get-your-api-key) to get your API key
  - Existing api keys can be found in the Alchemy Dashboard by viewing your config in the [Smart wallets configuration page](https://dashboard.alchemy.com/services/smart-wallets/configuration)
- Node ≥ 18 (or cURL / any HTTP client)
- The `@account-kit/smart-contracts` package (for step 4)

You can generate a wallet address on both EVM and Solana before a user ever authenticates or logs in using only their email. This could enable use cases such as:

- airdropping assets or seed balances to users who signed up for a waitlist with their email before they ever open your app
- allowing users to send assets to another user with just their email even if the recipient has not logged in before. When that recipient logs in, they'll be able to claim their balance.
- migrating assets to smart wallets to enable gas sponsorship, batching, etc.
- and more!

With Alchemy you can do this entirely from your server using the **Signer API** – no UI, hooks, or user interaction required.

> **Note:** Smart contract addresses are deterministic. The deployment address is a function of the address of owner/signer address, the account implementation (e.g. latest version of Modular Account), and the salt (optionally specified). If all three of those remain the same, then the smart account will be consistent across chains and for users. We will handle generating a consistent signer address for email.

Reviewer: ```suggestion

**Note:** Smart contract addresses are deterministic. The deployment address is a function of the address of owner/signer address, the account implementation (e.g. latest version of Modular Account), and the salt (optionally specified). If all three of those remain the same, then the smart account will be consistent across chains and for users. We will handle generating a consistent signer address for email.
```

---
title: Pregenerate Wallets
description: Learn how to pre-generate Smart Wallet addresses for your users with nothing more than an email address.
slug: wallets/react/pregenerate-wallets
---

> > **Objective** Generate a wallet owner (signer) and predict the smart-account address _before_ a user ever logs in, using only their email.

Reviewer: ```suggestion
> **Objective** Generate a wallet owner (signer) and the smart-account address _before_ a user ever logs in, using only their email.
```

## Prerequisites

> - Alchemy **API Key** with Smart Wallets enabled

Reviewer: Do you think these prereqs are relevant here? I feel like this applies to all docs and maybe we dont need to be repetitive here. We can just say something like - make sure you've set up your embedded wallets / enabled on your app. And link that to the quickstart/setup doc?

Author: Yeah I'll link to the quickstart instead

- Node ≥ 18 (or cURL / any HTTP client)
- The `@account-kit/smart-contracts` package (for step 4)

> When you already know who your users are (for example you collected their email during wait-list signup) you might want to **create wallet addresses ahead of time** so you can airdrop assets or seed balances _before_ they ever open your app.

Reviewer: ```suggestion
You can generate a wallet address on both EVM and Solana before a user ever authenticates or logs in using only their email. This could enable use cases such as: 
- airdropping assets or seed balances to users who signed up for a waitlist with their email before they ever open your app
- allowing users to send assets to another user with just their email even if the recipient has not logged in before. When that recipient logs in, they'll be able to claim their balance. 
- migrating assets to smart wallets to enable gas sponsorship, batching, etc.
- and more!
```

Love the idea of listing some example use cases. Added some more details but feel free to tweak.

With Alchemy you can do this entirely from your server using the **Signer API** – no UI, hooks, or user interaction required.

> 

Reviewer: ```suggestion
**Note:** Smart contract addresses are deterministic. The deployment address is a function of the address of owner/signer address, the account implementation (e.g. latest version of Modular Account), and the salt (optionally specified). If all three of those remain the same, then the smart account will be consistent across chains and for users. We will handle generating a consistent signer address for email.
```

> ## 1. Get an API key

Reviewer: Should this instead point to the setup doc - they need to enable wallets right? might not be worth re-explaining here

Grab an **API Key** from the Alchemy dashboard and store it as an environment variable:

```bash
export ALCHEMY_API_KEY="<YOUR_API_KEY>"
```

## 2. Create the owner signer

> Use whichever call style fits your environment – they all call the same [`Create Wallet` API endpoint](https://www.alchemy.com/docs/wallets/api/smart-wallets/signer-api-endpoints/signer-api-endpoints/create-account) with an identical JSON body.

Reviewer: ```suggestion
First, you need to generate the signer address. This will be generated using the [`Create Wallet` API endpoint](https://www.alchemy.com/docs/wallets/api/smart-wallets/signer-api-endpoints/signer-api-endpoints/create-account). The signer address will be used as the owner of the new smart account and enable the user to sign transactions after authenticating with their email.
```

<Tabs>
<Tab title="Fetch (ESM / Node ≥18)" language="ts">

```ts twoslash
const res = await fetch("https://api.g.alchemy.com/signer/v1/signup", {
  method: "POST",
  headers: {
    "Content-Type": "application/json",

>     // The API key is sent in the Authorization header

Reviewer: Note where to get api key

    Authorization: `Bearer ${process.env.ALCHEMY_API_KEY}`,
  },
  body: JSON.stringify({
    email: "[email protected]",
  }),
});

if (!res.ok) throw new Error(await res.text());

const { address, solanaAddress, userId } = await res.json();
console.log({ address, solanaAddress, userId });
```

</Tab>
<Tab title="request (CommonJS)" language="ts">

```ts
import request from "request";

request.post(
  {
    url: "https://api.g.alchemy.com/signer/v1/signup",
    json: { email: "[email protected]" },
    headers: {
      Authorization: `Bearer ${process.env.ALCHEMY_API_KEY}`,
      "Content-Type": "application/json",
    },
  },
  (err, _resp, body) => {
    if (err) throw err;
    console.log(body); // ➜  { orgId, userId, address, solanaAddress, otpId }
  }
);
```

</Tab>
<Tab title="cURL" language="bash">

```bash
curl -X POST https://api.g.alchemy.com/signer/v1/signup \
  -H "Authorization: Bearer $ALCHEMY_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{ "email": "[email protected]" }'
```

</Tab>
</Tabs>

> ## 3. Response fields

Reviewer: Is this a separete step or just part of 2

| Field           | Description                                                                                                        |
| --------------- | ------------------------------------------------------------------------------------------------------------------ |
| `address`       | The **signer (EOA)** address that will **own** your user's smart account (this is _not_ the smart account itself). |
| `solanaAddress` | The **Solana** wallet address associated with the same user.                                                       |
| `orgId`         | Your organisation ID – useful if you operate multiple orgs.                                                        |

<Tip>
  The address you receive here is the **owner key**. Your user's Smart Contract
  Account (ModularAccountV2) will be deployed at a different address – but we
  can deterministically calculate it ahead of time using
  `predictModularAccountV2Address`.
</Tip>

## 4. Predict the smart-account address

> The `@account-kit/smart-contracts` package exports a utility that mirrors the

Reviewer: _The `@account-kit/smart-contracts` package exports a utility that mirrors the
CREATE2 logic used by the MAv2 factory. By passing the signer address you just
generated, you can know the **counter-factual** (future) address of the smart
account before it is even deployed._

Would change this to:
Using the signer address you generated in step 2, you can now generate the **counter-factual** (future) address of the smart account before it is even deployed. 

The `predictModularAccountV2Address` method will generate the address for a new Modular Account v2 and mirrors the CREATE2 logic used by the MAv2 factory.
